tests/features/arg_matching.py

```diff
@@ -0,0 +1,255 @@
+"""Tests for argument matching."""
+# pylint: disable=missing-docstring,no-self-use,no-member
+import re
+
+from flexmock import exceptions, flexmock
+from flexmock._api import flexmock_teardown
+from tests import some_module
+from tests.some_module import SomeClass
+from tests.utils import assert_raises
+
+
+class ArgumentMatchingTestCase:
+    def test_arg_matching_works_with_regexp(self):
+        class Foo:
+            def method(self, arg1, arg2):
+                pass
+
+        instance = Foo()
+        flexmock(instance).should_receive("method").with_args(
+            re.compile("^arg1.*asdf$"), arg2=re.compile("f")
+        ).and_return("mocked")
+        assert instance.method("arg1somejunkasdf", arg2="aadsfdas") == "mocked"
+
+    def test_arg_matching_with_regexp_fails_when_regexp_doesnt_match_karg(self):
+        class Foo:
+            def method(self, arg1, arg2):
+                pass
+
+        instance = Foo()
+        flexmock(instance).should_receive("method").with_args(
+            re.compile("^arg1.*asdf$"), arg2=re.compile("a")
+        ).and_return("mocked")
+        with assert_raises(
+            exceptions.MethodSignatureError,
+            (
+                "Arguments for call method did not match expectations:\n"
+                '  Received call:\tmethod("arg1somejunkasdfa", arg2="a")\n'
+                "  Expected call[1]:\tmethod(arg2=/a/, arg1=/^arg1.*asdf$/)"
+            ),
+        ):
+            instance.method("arg1somejunkasdfa", arg2="a")
+
+    def test_arg_matching_with_regexp_fails_when_regexp_doesnt_match_kwarg(self):
+        class Foo:
+            def method(self, arg1, arg2):
+                pass
+
+        instance = Foo()
+        flexmock(instance).should_receive("method").with_args(
+            re.compile("^arg1.*asdf$"), arg2=re.compile("a")
+        ).and_return("mocked")
+        with assert_raises(
+            exceptions.MethodSignatureError,
+            (
+                "Arguments for call method did not match expectations:\n"
+                '  Received call:\tmethod("arg1somejunkasdf", arg2="b")\n'
+                "  Expected call[1]:\tmethod(arg2=/a/, arg1=/^arg1.*asdf$/)"
+            ),
+        ):
+            instance.method("arg1somejunkasdf", arg2="b")
+
+    def test_module_level_function_with_kwargs(self):
+        flexmock(some_module).should_receive("module_function").with_args(1, y="expected")
+        with assert_raises(
+            exceptions.FlexmockError,
+            (
+                "Arguments for call module_function did not match expectations:\n"
+                '  Received call:\tmodule_function(1, y="not expected")\n'
+                '  Expected call[1]:\tmodule_function(y="expected", x=1)'
+            ),
+        ):
+            some_module.module_function(1, y="not expected")
+
+    def test_flexmock_should_match_types_on_multiple_arguments(self):
+        class Foo:
+            def method1(self, arg1, arg2):
```

Comment:
Why is it `method1` now and `method` earlier ?

Reply:
It seems that it was named `method1` before my changes. I renamed the methods to `method` for consistency: https://github.com/flexmock/flexmock/pull/114/commits/c1e362b5a8319b8ca3dce1751e3f5180abfbdad0